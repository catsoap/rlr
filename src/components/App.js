import React, { Component } from 'react';
import axios from 'axios';
import Search from './Search';
import Table from './Table';
import { ButtonWithLoading } from './Button';

const DEFAULT_QUERY = 'cryptoeconomics';
const DEFAULT_HPP = '100';
const PATH_BASE = 'https://hn.algolia.com/api/v1';
const PATH_SEARCH = '/search';
const PARAM_SEARCH = 'query=';
const PARAM_PAGE = 'page=';
const PARAM_HPP = 'hitsPerPage=';

class App extends Component {
  _isMounted = false;

  constructor(props) {
    super(props);

    this.state = {
      results: null,
      searchKey: '',
      searchTerm: DEFAULT_QUERY,
      error: null,
      isLoading: false,
    };

    this.needsToSearchTopStories = this.needsToSearchTopStories.bind(this);
    this.setSearchTopStories = this.setSearchTopStories.bind(this);
    this.fetchSearchTopStories = this.fetchSearchTopStories.bind(this);
    this.onSearchChange = this.onSearchChange.bind(this);
    this.onSearchSubmit = this.onSearchSubmit.bind(this);
    this.onDismiss = this.onDismiss.bind(this);
  }

  needsToSearchTopStories(searchTerm) {
    return !this.state.results[searchTerm];
  }

  setSearchTopStories(result) {
    const { hits, page } = result;
    this.setState(updateSearchTopStoriesState(hits, page));
  }

  fetchSearchTopStories(searchTerm, page = 0) {
    this.setState({ isLoading: true });

    axios(
      `${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`,
    )
      .then((result) => {
        if (this._isMounted) {
          this.setSearchTopStories(result.data);
        }
      })
      .catch((error) => this.setState({ error }));
  }

  componentDidMount() {
    this._isMounted = true;
    const { searchTerm } = this.state;
    this.setState({ searchKey: searchTerm });
    this.fetchSearchTopStories(searchTerm);
  }

  componentWillUnmount() {
    this._isMounted = false;
  }

  onSearchChange(event) {
    this.setState({ searchTerm: event.target.value });
  }

  onSearchSubmit(event) {
    const { searchTerm } = this.state;
    this.setState({ searchKey: searchTerm });

    if (this.needsToSearchTopStories(searchTerm)) {
      this.fetchSearchTopStories(searchTerm);
    }

    event.preventDefault();
  }

  onDismiss(id) {
    const { searchKey, results } = this.state;
    const { hits, page } = results[searchKey];

    const isNotId = (item) => item.objectID !== id;
    const updatedHits = hits.filter(isNotId);

    this.setState({
      results: {
        ...results,
        [searchKey]: { hits: updatedHits, page },
      },
    });
  }

  render() {
    const { searchTerm, results, searchKey, error, isLoading } = this.state;
    const page = (results && results[searchKey] && results[searchKey].page) || 0;
    const list = (results && results[searchKey] && results[searchKey].hits) || [];

    return (
      <div>
        <div className="flex justify-center items-center flex-col md:flex-row my-12">
          <h1 className="md:order-1 text-lg text-gray-800 text-right px-4 mb-8 md:mb-0">
            Hackers News Search Engine
          </h1>
          <Search value={searchTerm} onChange={this.onSearchChange} onSubmit={this.onSearchSubmit}>
            Search
          </Search>
        </div>
        <div className="flex items-center flex-col">
          {error ? <p>Something went wrong.</p> : <Table list={list} onDismiss={this.onDismiss} />}
          {
            <div className="my-4">
              <ButtonWithLoading
                isLoading={isLoading}
                onClick={() => this.fetchSearchTopStories(searchKey, page + 1)}
                className="button"
              >
                More
              </ButtonWithLoading>
            </div>
          }
        </div>
      </div>
    );
  }
}

const updateSearchTopStoriesState = (hits, page) => (prevState) => {
  const { searchKey, results } = prevState;
  const oldHits = results && results[searchKey] ? results[searchKey].hits : [];
  const updatedHits = [...oldHits, ...hits];
  return {
    results: {
      ...results,
      [searchKey]: { hits: updatedHits, page },
    },
    isLoading: false,
  };
};

export default App;

export { updateSearchTopStoriesState };
