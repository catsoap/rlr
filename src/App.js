import React, { useEffect, useState, useReducer } from 'react';
import axios from 'axios';
import { Box, Container, ThemeProvider } from 'theme-ui';
import theme from './theme';
import Header from './components/Header';
import Search from './components/Search';
import Table from './components/Table';
import { ButtonWithLoading } from './components/Button';

const DEFAULT_QUERY = 'react';
const DEFAULT_HPP = '100';

function App() {
  const [query, setQuery] = useState('redux');
  const [{ data, isLoading, isError }, doFetch] = useDataApi(
    `https://hn.algolia.com/api/v1/search?query=${DEFAULT_QUERY}&page=0&hitsPerPage=${DEFAULT_HPP}`,
    {
      hits: [],
    },
  );

  const { page, nbPages } = data;

  return (
    <ThemeProvider theme={theme}>
      <Container p={1}>
        <Header>
          <Search
            value={query}
            onChange={(e) => setQuery(e.target.value)}
            onSubmit={(event) => {
              doFetch(
                `http://hn.algolia.com/api/v1/search?query=${query}&page=0&hitsPerPage=${DEFAULT_HPP}`,
              );
              event.preventDefault();
            }}
          >
            Search
          </Search>
        </Header>
        <div>
          {isError ? <p>Something went wrong.</p> : <Table list={data.hits} />}
          <Box sx={{ textAlign: 'center', my: 2 }}>
            {(isLoading || !page < nbPages) && (
              <ButtonWithLoading
                isLoading={isLoading}
                onClick={() => {
                  doFetch(
                    `http://hn.algolia.com/api/v1/search?query=${query}&page=${
                      page + 1
                    }&hitsPerPage=${DEFAULT_HPP}`,
                  );
                }}
                className="button"
              >
                More
              </ButtonWithLoading>
            )}
          </Box>
        </div>
      </Container>
    </ThemeProvider>
  );
}

const dataFetchReducer = (state, action) => {
  switch (action.type) {
    case 'FETCH_INIT': {
      const data = action.refresh ? { hits: [] } : state.data;

      return {
        ...state,
        data,
        isLoading: true,
        isError: false,
      };
    }
    case 'FETCH_SUCCESS': {
      let data;
      const { payload } = action;
      if (payload.page === 0) {
        data = payload;
      } else {
        const updatedHits = [...state.data.hits, ...payload.hits];
        data = { ...payload, ...{ hits: updatedHits } };
      }

      return {
        ...state,
        data,
        isLoading: false,
        isError: false,
      };
    }
    case 'FETCH_FAILURE': {
      return {
        ...state,
        isLoading: false,
        isError: true,
      };
    }
    default:
      throw new Error();
  }
};

const useDataApi = (initialUrl, initialData) => {
  const [url, setUrl] = useState(initialUrl);
  const [state, dispatch] = useReducer(dataFetchReducer, {
    isLoading: false,
    isError: false,
    data: initialData,
  });
  useEffect(() => {
    let didCancel = false;
    const fetchData = async () => {
      dispatch({ type: 'FETCH_INIT', refresh: -1 !== url.indexOf('page=0') });
      try {
        const result = await axios(url);
        if (!didCancel) {
          dispatch({ type: 'FETCH_SUCCESS', payload: result.data });
        }
      } catch (error) {
        if (!didCancel) {
          dispatch({ type: 'FETCH_FAILURE' });
        }
      }
    };
    fetchData();
    return () => {
      didCancel = true;
    };
  }, [url]);
  return [state, setUrl];
};

export default App;
