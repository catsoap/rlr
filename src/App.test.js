import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import App, { updateSearchTopStoriesState } from './App';

describe('App', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test('has a valid snapshot', () => {
    const component = renderer.create(<App />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Search Top Stories State', () => {
  it('returns a new state', () => {
    const prevState = {
      results: null,
      searchKey: 'redux',
      searchTerm: 'redux',
      error: null,
      isLoading: true,
    };

    const hits = [{ title: '1', author: '1', num_comments: 1, points: 2, objectID: 'x' }];
    const page = 1;
    const newState = updateSearchTopStoriesState(hits, page)(prevState);
    expect(newState.results['redux'].hits).toMatchObject(hits);
    expect(newState.results['redux'].page).toBe(page);
  });
});
