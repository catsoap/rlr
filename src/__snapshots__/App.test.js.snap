// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App has a valid snapshot 1`] = `
<div
  className="css-6nxnnj"
>
  <header
    className="css-118d9or-Header"
  >
    <h1
      className="css-bwl8bm-Header"
    >
      Hackers News Search Engine
    </h1>
    <form
      className="css-1lchotj"
      onSubmit={[Function]}
    >
      <svg
        className="css-qdzqyl"
      >
        search.svg
      </svg>
      <input
        className="css-119tjkq"
        onChange={[Function]}
        value="react"
      />
      <button
        className="css-1cenlpe"
        type="submit"
      >
        <span
          className="css-1hy7ipv"
        >
          Search
        </span>
        <svg
          className="css-lgquwi"
        >
          search.svg
        </svg>
      </button>
    </form>
  </header>
  <div>
    <section>
      <header
        className="css-17n9c24"
      >
        <div
          className="css-mu2wql"
        >
          <button
            className="css-7cs3ym-Sort"
            onClick={[Function]}
          >
            Title
          </button>
        </div>
        <div
          className="css-ieri5n"
        >
          <button
            className="css-7cs3ym-Sort"
            onClick={[Function]}
          >
            Author
          </button>
        </div>
        <div
          className="css-1rr4qq7"
        >
          <button
            className="css-7cs3ym-Sort"
            onClick={[Function]}
          >
            Coms
          </button>
        </div>
        <div
          className="css-1rr4qq7"
        >
          <button
            className="css-7cs3ym-Sort"
            onClick={[Function]}
          >
            Pts
          </button>
        </div>
        <div
          className="css-kzfr2u"
        >
          <button
            className="css-7cs3ym-Sort"
            onClick={[Function]}
          >
            Date
          </button>
        </div>
        <div
          className="css-1rr4qq7"
        />
      </header>
    </section>
    <div
      className="css-1drvtrq"
    >
      <svg
        className="css-sqxpzz"
        fill="none"
        height={48}
        role="img"
        stroke="currentcolor"
        strokeWidth={4}
        viewBox="0 0 32 32"
        width={48}
      >
        <title>
          Loading...
        </title>
        <circle
          cx={16}
          cy={16}
          opacity={0.125}
          r={12}
        />
        <circle
          className="css-wpcq6n"
          cx={16}
          cy={16}
          r={12}
          strokeDasharray={75.39822368615503}
          strokeDashoffset={56.548667764616276}
        />
      </svg>
    </div>
  </div>
</div>
`;
